typescript/infra/scripts/verify-validator.ts

```diff
@@ -0,0 +1,297 @@
+  readonly bucket: string;
+
+  constructor(bucketUrl: string) {
+    const match = bucketUrl.match(
+      /^(?:https?:\/\/)?(.*)\.s3\.(.*)\.amazonaws.com\/?$/,
+    );
+    if (!match) throw new Error('Could not parse bucket url');
+    this.bucket = match[1];
+    this.region = match[2];
+    this.client = new S3Client({ region: this.region });
+  }
+
+  async getS3Obj<T = unknown>(
+    key: string,
+  ): Promise<{ obj: T; modified: Date }> {
+    const response = await this.client.send(
+      new GetObjectCommand({
+        Bucket: this.bucket,
+        Key: key,
+      }),
+    );
+    if (!response.Body) {
+      throw new Error('No data received');
+    }
+    const bodyStream: NodeJS.ReadableStream =
+      'stream' in response.Body
+        ? response.Body.stream()
+        : (response.Body as NodeJS.ReadableStream);
+
+    const body: string = await streamToString(bodyStream);
+    return {
+      obj: JSON.parse(body),
+      modified: response.LastModified!,
+    };
+  }
+}
+
+async function main() {
+  const {
+    a: _validatorAddress,
+    p: prospectiveBucket,
+    c: controlBucket,
+  } = await getArgs();
+
+  const cClient = new S3Wrapper(controlBucket);
+  const pClient = new S3Wrapper(prospectiveBucket);
+
+  const [{ obj: cLatestCheckpoint }, { obj: pLastCheckpoint }] =
+    await Promise.all([
+      cClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+        console.error(
+          "Failed to get control validator's latest checkpoint.",
+          err,
+        );
+        process.exit(1);
+      }),
+      pClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+        console.error(
+          "Failed to get prospective validator's latest checkpoint.",
+          err,
+        );
+        process.exit(1);
+      }),
+    ]);
+
+  console.assert(
+    Number.isSafeInteger(cLatestCheckpoint),
+    'Expected latest control checkpoint to be an integer',
+  );
+  console.assert(
+    Number.isSafeInteger(pLastCheckpoint),
+    'Expected latest prospective checkpoint to be an integer',
+  );
+
+  console.log(`Latest Index`);
+  console.log(`control: ${cLatestCheckpoint}`);
+  console.log(`prospective: ${pLastCheckpoint}\n`);
+
+  let extraCheckpoints = [];
+  const missingCheckpoints = [];
+  let invalidCheckpoints = [];
+  const modTimeDeltasS = [];
+  const fullyCorrectCheckpoints = [];
+  let missingInARow = 0;
+  let lastNonMissingCheckpointIndex = Infinity;
+  for (let i = Math.max(cLatestCheckpoint, pLastCheckpoint); i >= 0; --i) {
+    if (missingInARow == MAX_MISSING_CHECKPOINTS) {
+      missingCheckpoints.length -= MAX_MISSING_CHECKPOINTS;
+      invalidCheckpoints = invalidCheckpoints.filter(
+        (j) => j < lastNonMissingCheckpointIndex,
+      );
+      extraCheckpoints = extraCheckpoints.filter(
+        (j) => j < lastNonMissingCheckpointIndex,
+      );
+      break;
+    }
+
+    const key = `checkpoint_${i}.json`;
+
+    let c: Checkpoint | null;
```

Comment:
also noticing the single letter `c` here and `cLastMod` is a little confusing bc of how it's also used for `control`